docs/pallas/tpu/matmul.md

```diff
@@ -0,0 +1,646 @@
+
+The minimum amount of memory usage for a matrix multiply (assuming float32) is the total size of the inputs (copying into VMEM) plus the size of the output (copying into HBM). Thus the minimum bandwidth usage is `(m * k + k * n + m * n) * 4 bytes/float32`.
+
+One observation is that the number of matmul FLOPs is cubic in its inputs whereas the minimum bandwidth usage is quadratic in its inputs. Intuitively, this means that FLOPs grow faster than bandwidth usage, meaning that the bigger our matmul is, the more compute we have relative to copying.
+
+```{code-cell}
+:id: HZwmYZ61QZ5L
+:outputId: 18505741-9254-4738-ec64-1660f6733d77
+
+def matmul_flops(m: int, k: int, n: int):
+  return 2 * m * k * n
+
+def matmul_membw(m: int, k: int, n: int, dtype: jnp.dtype):
+  return (m * k + k * n + m * n) * np.dtype(dtype).itemsize
+
+print(matmul_flops(1024, 1024, 1024))
+print(matmul_membw(1024, 1024, 1024, jnp.float32))
+```
+
++++ {"id": "agCtb2GMQazl"}
+
+Now that we can calculate the amount of FLOPs and memory bandwidth usage of a matrix multiplication, let's now see how many FLOP/s we can execute and how much memory bandwidth we actually have on a particular TPU.
+
+This notebook was run on TPU v5e so we will utilize numbers from there (if you are running this notebook, your numbers may differ). TPU v5es have [197 TFLOPs of bf16 compute and 819 GBps of memory bandwidth](https://cloud.google.com/tpu/docs/system-architecture-tpu-vm#tpu_v5e).
+
+```{code-cell}
+:id: WUydNX2-K6Oy
+
+v5e_bf16_flops = 197e12
+v5e_membw = 819e9
+```
+
++++ {"id": "UjQIWq-9RJue"}
+
+We can now estimate the amount of time it takes to execute a matmul of a particular size using the FLOPs and how much time it takes to copy the inputs and outputs using the memory bandwidth.
+
+```{code-cell}
+:id: PiYobLc-RQSB
+
+def matmul_flops_time_us(m: int, k: int, n: int):
+  flops = matmul_flops(m, k, n)
+  return flops / v5e_bf16_flops * 10e6
+
+def matmul_copy_time_us(m: int, k: int, n: int, dtype: jnp.dtype):
+  membw = matmul_membw(m, k, n, dtype)
+  return membw / v5e_membw * 10e6
+```
+
++++ {"id": "q1y6dP00Sv9S"}
+
+This basic calculation tells us roughly how efficiently we'll be able to use our MXUs. If it will take more time to copy our inputs/outputs to and from HBM than it will to execute the FLOPs, then our computation will be *memory bound*, i.e. our compute units will be idling while waiting for values to be transferred. If we take more time to matmul than to copy, then we are *compute bound*, then our compute units will never be idle and we will be *compute bound*.
+
+Because matmul flops are cubic in their input sizes and memory bandwidth usage is quadratic, we expect that we will get compute bound as we get bigger and bigger, but this crossing over point is really important! Let's say we are doing a `(1024, 1024) x (1024, 1024)` float32 matrix multiplication (`f32`s get truncated to `bf16` in the MXU so we use the `bf16` flops numbers but `f32` memory bandwidth numbers).
```

Comment:
```suggestion
Because matmul FLOPs are cubic in their input sizes and memory bandwidth usage is quadratic, we expect that we will get compute bound as we get bigger and bigger, but this crossing over point is really important! Let's say we are doing a `(1024, 1024) x (1024, 1024)` float32 matrix multiplication (`f32`s get truncated to `bf16` in the MXU so we use the `bf16` flops numbers but `f32` memory bandwidth numbers).
```